### ISSUE TITLE
[BUG] Inconsistent tensor shapes and zero probability handling in multinomial sampling

### ISSUE THREAD

**Issue by eigenvalnull** [synnada-ai/mithril#7]
# Bug Report

### Describe the Bug
inconsistent behavior in multinomial sampling between torch/jax backends. 

### To Reproduce
```python
import mithril as ml

# init backends
torch_backend = ml.TorchBackend()
jax_backend = ml.JaxBackend()

# test basic sampling
probs = [[0.7, 0.3], [0.4, 0.6]]
torch_probs = torch_backend.array(probs) 
jax_probs = jax_backend.array(probs)

torch_samples = torch_backend.multinomial(torch_probs, num_samples=1, replacement=False)
jax_samples = jax_backend.multinomial(jax_probs, num_samples=1, replacement=False)

print(f"torch shape: {torch_samples.shape}")  # torch.Size([2, 1])
print(f"jax shape: {jax_samples.shape}")     # (2, 1, 1)

# gets worse w/ zero probs
edge_probs = [[0.0, 1.0], [0.5, 0.5]]
torch_probs = torch_backend.array(edge_probs)
jax_probs = jax_backend.array(edge_probs)

torch_samples = torch_backend.multinomial(torch_probs, num_samples=1, replacement=False)
jax_samples = jax_backend.multinomial(jax_probs, num_samples=1, replacement=False) 

print("torch:", torch_samples)  # tensor([[1], [1]])
print("jax:", jax_samples)     # [[[1]] [[0]]]

```

### Expected Behavior

- consistent tensor shapes between backends (probably follow torch's 2d convention?)
- same sampling behavior with edge cases like zero probabilities

### Additional Context
saw @ozankabak's talk about mithril at cmu db seminar (btw - is this meant to compete w/ keras or more compiler/IR focused?). can help fix if helpful, p.s. the parallelization stuff looks super cool, gonna try that next.

### System Info

**OS**: macOS Sonoma
**Mithril Version**: 0.1.1
**Torch Version**: 2.4.x
**MLX Version**: 0.19.3
**NumPy Version**: 2.1.x
**JAX Version**: 0.4.35
**Python Version**: 3.12.7

**Comment by ozankabak** [comment]
You are right, shapes should match. I am not clear on the second issue, though. These are sampling functions, they are supposed to generate random results based on the given probabilities. Do you expect the results of the first call to be the same with those of the second?

BTW, we are indeed more compiler/IR focused. Keras-like functionality emerges as a corollary of that (if we do a good job) 🙂

**Comment by ozankabak** [comment]
Thank you for the clarification - we will take a look and circle back soon.

Also, please keep the feedback and contributions coming. We are excited to collaborate with everyone who thinks this kind of an approach is promising (and fun)!

**Comment by mehmetozsoy-synnada** [comment]
I reviewed your test cases and example. Perhaps I misunderstood something, but I couldn't reproduce the issue you described.

Here is the code I wrote to try reproducing the problem. I used the numbers from your `test_edge_cases` test, as I assumed it is the test intended to cover the issue. (By the way, in these tests, `num_samples` was higher than the number of probabilities, which causes an error when `replacement=False` by default. I assumed `replacement=True`?)

```python
import mithril as ml
backend_torch = ml.TorchBackend(precision=32)
backend_jax = ml.JaxBackend(precision=32)
n_samples = 1000

# Probabilities in Torch
probs_torch = backend_torch.array([0.0, 0.5, 0.5])
samples_torch = backend_torch.multinomial(probs_torch, num_samples=n_samples, replacement=True)
2 in samples_torch  # True
1 in samples_torch  # True
0 in samples_torch  # False

# Probabilities in Jax
probs_jax = backend_jax.array([0.0, 0.5, 0.5])
samples_jax = backend_jax.multinomial(probs_jax, num_samples=n_samples, replacement=True)
2 in samples_jax  # True
1 in samples_jax  # True
0 in samples_jax  # False
```
In this code, `0` is never sampled in both Jax and Torch.

I also examined the case of a `[[0.0, 1.0], [0.5, 0.5]]` distribution. This setup contains two distributions, producing an output with two numbers. The first element in the output is sampled from the `[0.0, 1.0]` distribution. Thus, I wouldn’t expect the first item of the output to be `0`. However, the second number in the output is sampled from the `[0.5, 0.5]` distribution, so it could be either `0` or `1`.

Therefore, apart from the shape mismatch, the outputs `[[[1], [0]]]` in Jax and `[[1], [1]]` in Torch seem correct to me. Please let me know if I am missing any point.

**Comment by mehmetozsoy-synnada** [comment]
After examining the multinomial code a bit further, you were right. In Mithril's JAX implementation of multinomial, we replace the zero-probability case with `1e-37` to avoid numerical instabilities. This means that, with a small probability, zero can still be sampled in JAX. Thank you for pointing it out. We would be excited to see any contributions toward addressing this issue.

**Comment by ozankabak** [comment]
AFAICT @mehmetozsoy-synnada's point on `replacement` still stands, but there indeed seems to be an issue w.r.t. how edge cases are handled in JAX. If you don't have the time to contribute a PR, we will fix this in the next month or so, after we tackle the higher priority issues before the GA release of Mithril.

**Comment by SinanGncgl** [comment]
Sorry for bringing this up again. I wanted to investigate and address the issue, but it seems the problem is no longer valid. After following the reproduction steps, I’m getting consistent results.

**Comment by ozankabak** [comment]
@mehmetozsoy-synnada, did we fix this yet?

**Comment by mehmetozsoy-synnada** [comment]
The inconsistent tensor shapes bug in JAX has been fixed and merged into the main branch at #145. However, the zero-probability edge case handling bug in JAX still persists in the main branch.

**Comment by ozankabak** [comment]
@SinanGncgl, if you want to fix the edge case behavior involving unnecessary epsilon usage, we would appreciate it 🚀  We can also fix it on our end if you can't find the time.

**Comment by SinanGncgl** [comment]
I’ve submitted a PR to address the issue. Let me know if you need any clarifications or would like to suggest any changes.

**Comment by ozankabak** [comment]
I think this issue is fixed now. @eigenvalnull, if we are mistaken, feel free to reopen and let us know. Thanks a lot for reporting this issue.
